graph/db/sql_store.go

```diff
@@ -2215,6 +2223,206 @@ func (s *SQLStore) PruneGraphNodes() ([]route.Vertex, error) {
 	return prunedNodes, nil
 }
 
+// PruneGraph prunes newly closed channels from the channel graph in response
+// to a new block being solved on the network. Any transactions which spend the
+// funding output of any known channels within he graph will be deleted.
+// Additionally, the "prune tip", or the last block which has been used to
+// prune the graph is stored so callers can ensure the graph is fully in sync
+// with the current UTXO state. A slice of channels that have been closed by
+// the target block along with any pruned nodes are returned if the function
+// succeeds without error.
+//
+// NOTE: part of the V1Store interface.
+func (s *SQLStore) PruneGraph(spentOutputs []*wire.OutPoint,
+	blockHash *chainhash.Hash, blockHeight uint32) (
+	[]*models.ChannelEdgeInfo, []route.Vertex, error) {
+
+	ctx := context.TODO()
+
+	s.cacheMu.Lock()
+	defer s.cacheMu.Unlock()
+
+	var (
+		closedChans []*models.ChannelEdgeInfo
+		prunedNodes []route.Vertex
+	)
+	err := s.db.ExecTx(ctx, sqldb.WriteTxOpt(), func(db SQLQueries) error {
+		for _, outpoint := range spentOutputs {
+			row, err := db.GetChannelByOutpoint(
```

Comment:
I think we should add a TODO here for optimizing this. On mainnet this will potentially create thousands of SQL queries (one for each spent input in all transactions of a block).
Unfortunately there is no easy way to use `WHERE c.outpoint IN (@outpoint_list)` that works for both SQLite and Postgres. I don't remember what exactly, but something with [`sqlc.slice()`](https://docs.sqlc.dev/en/latest/reference/macros.html#sqlc-slice) didn't work properly because we're using the SQLite engine in `sqlc`. But perhaps worth a try with the latest `sqlc` version?

Can also be deferred to a later performance optimizing PR, so not a blocker here.